packages/matter.js/src/codec/DerCodec.ts

```diff
@@ -68,89 +83,155 @@ export class DerCodec {
+                if (value[TYPE_OVERRIDE_KEY] === DerType.Integer && value[RAW_DATA_KEY] instanceof ByteArray) {
+                    return this.encodeInteger(value[RAW_DATA_KEY]);
+                } else if (value[TYPE_OVERRIDE_KEY] === DerType.BitString && typeof value[RAW_DATA_KEY] === "number") {
+                    return this.encodeBitString(value[RAW_DATA_KEY]);
+                } else if (
+                    value[TYPE_OVERRIDE_KEY] === DerType.PrintableString &&
+                    typeof value[RAW_DATA_KEY] === "string"
+                ) {
+                    return this.encodePrintableString(value[RAW_DATA_KEY]);
+                } else if (value[TYPE_OVERRIDE_KEY] === DerType.IA5String && typeof value[RAW_DATA_KEY] === "string") {
+                    return this.encodeIA5String(value[RAW_DATA_KEY]);
+                } else {
+                    throw new UnexpectedDataError(`Unsupported override type ${value[TYPE_OVERRIDE_KEY]}`);
+                }
+            } else if (
+                value[BYTES_KEY] !== undefined &&
+                value[BYTES_KEY] instanceof ByteArray &&
+                Object.keys(value).length === 1
+            ) {
+                // Raw Data
+                return value[BYTES_KEY];
+            } else if (value[TYPE_OVERRIDE_KEY] === undefined && value[BYTES_KEY] === undefined) {
+                return this.encodeObject(value);
+            } else {
+                throw new UnexpectedDataError(`Unsupported object type ${typeof value}`);
+            }
         } else {
             throw new UnexpectedDataError(`Unsupported type ${typeof value}`);
         }
     }
 
     private static encodeDate(date: Date) {
-        return this.encodeAnsi1(
-            DerType.UtcDate,
-            ByteArray.fromString(
-                date
-                    .toISOString()
-                    .replace(/[-:.T]/g, "")
-                    .slice(2, 14) + "Z",
-            ),
-        );
+        if (date.getFullYear() > 2049) {
+            // Dates 2050+ are encoded as GeneralizedTime. This includes the special Non Well Defined date 9999-12-31.
+            return this.encodeAns1(
+                DerType.GeneralizedTime,
+                ByteArray.fromString(
+                    date
+                        .toISOString()
+                        .replace(/[-:.T]/g, "")
+                        .slice(0, 14) + "Z",
+                ),
+            );
+        } else
+            return this.encodeAns1(
+                DerType.UtcDate,
+                ByteArray.fromString(
+                    date
+                        .toISOString()
+                        .replace(/[-:.T]/g, "")
+                        .slice(2, 14) + "Z",
+                ),
+            );
     }
 
     private static encodeBoolean(bool: boolean) {
-        return this.encodeAnsi1(DerType.Boolean, ByteArray.of(bool ? 0xff : 0));
+        return this.encodeAns1(DerType.Boolean, ByteArray.of(bool ? 0xff : 0));
```

Comment:
Need to swap "n" and "s" too 🙂... Am on mobile at the moment or I'd do it

Reply:
Oohhh. Lol. Missed that. I do